cub/device/dispatch/dispatch_batch_memcpy.cuh

```diff
@@ -0,0 +1,631 @@
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ ******************************************************************************/
+
+/**
+ * \file
+ * cub::DispatchBatchMemcpy provides device-wide, parallel operations for copying data from a number of given source
+ * buffers to their corresponding destination buffer.
+ */
+
+#pragma once
+
+/// Optional outer namespace(s)
+#include <cstdint>
+
+#include <thrust/system/cuda/detail/core/triple_chevron_launch.h>
+
+#include "../../agent/agent_batch_memcpy.cuh"
+#include "../../agent/single_pass_scan_operators.cuh"
+#include "../../config.cuh"
+#include "../../thread/thread_search.cuh"
+
+CUB_NAMESPACE_BEGIN
+
+/**
+ * Initialization kernel for tile status initialization (multi-block)
+ */
+template <typename ScanTileStateT, typename TileOffsetT>
+__global__ void InitTileStateKernel(ScanTileStateT tile_state, TileOffsetT num_tiles)
+{
+  // Initialize tile status
+  tile_state.InitializeStatus(num_tiles);
+}
+
+/**
+ * Parameterizable tuning policy type for AgentBatchMemcpy
+ */
+template <uint32_t _BLOCK_THREADS, uint32_t _BYTES_PER_THREAD>
+struct AgentBatchMemcpyLargeBuffersPolicy
+{
+  enum
+  {
+    /// Threads per thread block
+    BLOCK_THREADS = _BLOCK_THREADS,
+    /// The number of bytes each thread copies
+    BYTES_PER_THREAD = _BYTES_PER_THREAD
+  };
+};
+
+/**
+ * Kernel that copies buffers that need to be copied by at least one (and potentially many) thread blocks.
+ */
+template <typename ChainedPolicyT,
+          typename BufferOffsetT,
+          typename InputBufferIt,
+          typename OutputBufferIt,
+          typename BufferSizeIteratorT,
+          typename BufferTileOffsetItT,
+          typename TileT,
+          typename TileOffsetT>
+__launch_bounds__(int(ChainedPolicyT::ActivePolicy::AgentLargeBufferPolicyT::BLOCK_THREADS)) __global__
+  void MultiBlockBatchMemcpyKernel(InputBufferIt input_buffer_it,
+                                   OutputBufferIt output_buffer_it,
+                                   BufferSizeIteratorT buffer_sizes,
+                                   BufferTileOffsetItT buffer_tile_offsets,
+                                   TileT buffer_offset_tile,
+                                   TileOffsetT last_tile_offset)
+{
+  using StatusWord    = typename TileT::StatusWord;
+  using ActivePolicyT = typename ChainedPolicyT::ActivePolicy::AgentLargeBufferPolicyT;
+
+  constexpr uint32_t BLOCK_THREADS    = ActivePolicyT::BLOCK_THREADS;
+  constexpr uint32_t ITEMS_PER_THREAD = ActivePolicyT::BYTES_PER_THREAD;
+  constexpr uint32_t ITEMS_PER_BLOCK  = BLOCK_THREADS * ITEMS_PER_THREAD;
+
+  __shared__ BufferOffsetT num_blev_buffers;
+
+  uint32_t tile_id = blockIdx.x;
+  if (threadIdx.x < CUB_PTX_WARP_THREADS)
+  {
+    StatusWord status_word;
+    BufferOffsetT tmp_blev_buffers;
+    buffer_offset_tile.WaitForValid(last_tile_offset, status_word, tmp_blev_buffers);
```

Comment:
This kernel is launched in a stream order. It's guaranteed to start executing after the previous one completes. Are there any reasons to over complicate this? I suggest we extend `ScanTileState` interface with the following methods:
```cpp
__device__ __forceinline__ T LoadValid(int tile_idx)                        
{                                                                           
    return d_tile_inclusive[TILE_STATUS_PADDING + tile_idx];
}
  
__device__ __forceinline__ T LoadValid(int tile_idx)                        
{                                                                           
    TxnWord alias = d_tile_descriptors[TILE_STATUS_PADDING + tile_idx];
    TileDescriptor tile_descriptor = reinterpret_cast<TileDescriptor&>(alias);
    return tile_descriptor.value;                                           
}  
```

In this case we'll spare some shared memory, a few CTA syncs, and enjoy reading `num_blev_buffers` from cache.

Reply:
Thanks for the suggestion! This was indeed just to work around the lack of such an interface in the tile state. I've adopted your suggestion 👍